adsft/tasks.py

```diff
@@ -190,6 +197,39 @@ def task_identify_facilities(message):
 
         writer.write_file(output_file_path, out)
 
+@app.task(queue='nlp-techniques')
+def task_apply_nlp_technqiues(message):
+
+    if not isinstance(message, list):
+        message = [message]
+
+    content = []
+    for m in message:
+        meta = checker.load_meta_file(m, app.conf['FULLTEXT_EXTRACT_PATH'])
+        ft = reader.read_content(meta)
+        if ft is not None:
+            content.append(ft)
+
+    for r in content:
+
+        bibcode_pair_tree_path = os.path.dirname(r['meta_path'])
+        output_file_path = os.path.join(bibcode_pair_tree_path, 'nlp.json')
+
+        doc = en_model(r['fulltext'])
+        out = doc.to_json()
+        out['ents'] = [{'text':ent.text, 'label':ent.label_} for ent in doc.ents]
+        out['noun-phrases'] = [chunk.text for chunk in doc.noun_chunks]
+        out['tokens'] = []
+        for tok in doc:
+            if not tok.is_punct and tok.is_alpha and not tok.is_stop and len(tok.text)>2:
+                out['tokens'].append({'token':tok.lower_,
+                                           'pos':tok.pos_,
+                                           'tag':tok.tag_,
+                                           'dep':tok.dep_,
+                                           'lemma':tok.lemma_})
```

Comment:
- Yes, we discussed how the text should be stored in a previous meeting in which the option was either to store the location that then corresponds to the fulltext file, or the actual entities. You said I could decide, and I chose to store the actual text knowing that there is a chance the fulltext and the nlp json file can get out of sync given the added config variables. Not having the context is a weakness of this output style. 
- It does have sentence tokenization, but in this case, only the location of the sentences are stored. 
- See the first bullet point. 
-  Normally in natural language processing you would choose either the lemma or the stem, this is something that I was told in my NLP course. Stemming is full of more errors so I chose lemmatization.
- You're right, I missed removing accents. The condition for `tok.is_alpha` should be removing all non-alphanumeric characters as instructed, but we can talk more about modifications on the preprocessing steps in today's meeting.
- "AB stars" would not be considered a token, "AB" and "stars" would be the tokens produced, and "AB" would be removed/ignored due to that condition. Keep in mind the noun-phrases do not have this same condition applied.  
- In our meeting, we discussed having the spell checker be a separate investigation, that could then be applied if it proves to be useful. You said it did not have to be part of this PR, but it is part of the project as a whole. "Only evaluate arXiv, if no time to incorporate into pipeline we can keep out of scope"

Reply:
* As discussed in person, let's consider saving the spacy object to keep the context
* Understood, it is in the json that comes from the spacy object
* Ok
* I understand lemmatization is generally prefered, my thought is that stemming is cheaper so if we can store both we will be able to validate in the future if stemming would be enough for our purposes or we do really need lemmas
* Let's improve the normalization then homogenizing accents and special characters
* My whole point is that we miss "AB" as a token, so I am looking for the justification of that condition (i.e., in what scenarios it is more beneficial to have it than not have it)
* It is not a priority to incorporate it into the pipeline (we agreed that's out of scope), but the analysis needs to be done with a conclusion recommending to include it or not in the future

Reply:
Regarding storing the spacy object (I'm all the time assuming we will stick to spaCy, that's up to your analysis), you mentioned by private communication that it would need 4.5MB instead of 59KB. To adjust to our current space capacity, you could reduce the first approximation to the vocabulary construction to articles that match this query: `((bibstem:(ApJ) collection:astronomy) AND year:2015-2020)`. This represents 16,136 entries, which translates to about 66GB of stored data.

Reply:
I'll add stemming into the script for preprocessing that now occurs outside of the pipeline. All tokens are now in Unicode to take care of accents/special characters. I am planning on completing the spellcheck analysis by the deadline, does it need to be completed before this PR is closed?

I made a colab [here](https://drive.google.com/drive/u/1/folders/1DYZHoSbv6ZR6WSmQ8huBIckF7qa97MaE) that I was going to use to show the differences in preprocessing. The reason I removed tokens<2chars is because it cleaned up tokens that I didn't want in my output. I was taking a more aggressive approach knowing the corpus vocabulary/dictionary needed to be kept at a reasonable size, and that is the only specific future task I know of to base these preprocessing decisions around. Preprocessing steps are supposed to change according to the task, so from my perspective, this should be more of a discussion because you may have future tasks in mind that I don't. For example, I didn't remove tokens<2chars when analyzing the XML parsers because I wanted to see the difference in the parsers even if it was minimal.   

I already wrote up some code to try out the language detection, but the issue I mentioned with celery also occurs with all of the non-English language models. I see that you have commented on some possible solutions, but as of right now I am blocked from implementing this additional feature. I was planning to create a dictionary of the models and use the output of [langdetect](https://github.com/Mimino666/langdetect) as the key, see all non-English models below:  

``` 
downloading the models:
https://github.com/explosion/spacy-models/releases/download/de_core_news_sm-2.2.5/de_core_news_sm-2.2.5.tar.gz
https://github.com/explosion/spacy-models/releases/download/fr_core_news_sm-2.2.5/fr_core_news_sm-2.2.5.tar.gz
https://github.com/explosion/spacy-models/releases/download/es_core_news_sm-2.2.5/es_core_news_sm-2.2.5.tar.gz
https://github.com/explosion/spacy-models/releases/download/pt_core_news_sm-2.2.5/pt_core_news_sm-2.2.5.tar.gz
https://github.com/explosion/spacy-models/releases/download/it_core_news_sm-2.2.5/it_core_news_sm-2.2.5.tar.gz
https://github.com/explosion/spacy-models/releases/download/nl_core_news_sm-2.2.5/nl_core_news_sm-2.2.5.tar.gz
https://github.com/explosion/spacy-models/releases/download/el_core_news_sm-2.2.5/el_core_news_sm-2.2.5.tar.gz
https://github.com/explosion/spacy-models/releases/download/nb_core_news_sm-2.2.5/nb_core_news_sm-2.2.5.tar.gz
https://github.com/explosion/spacy-models/releases/download/lt_core_news_sm-2.2.5/lt_core_news_sm-2.2.5.tar.gz

config:
DE_MODEL = 'de_core_news_sm'
FR_MODEL = 'fr_core_news_sm'
ES_MODEL = 'es_core_news_sm'
PT_MODEL = 'pt_core_news_sm'
IT_MODEL = 'it_core_news_sm'
NL_MODEL = 'nl_core_news_sm'
EL_MODEL = 'el_core_news_sm'
NB_MODEL = 'nb_core_news_sm'
LT_MODEL = 'lt_core_news_sm'

loading the models:
de_model = spacy.load(app.conf['DE_MODEL'])
fr_model = spacy.load(app.conf['FR_MODEL'])
es_model = spacy.load(app.conf['ES_MODEL'])
pt_model = spacy.load(app.conf['PT_MODEL'])
it_model = spacy.load(app.conf['IT_MODEL'])
nl_model = spacy.load(app.conf['NL_MODEL'])
el_model = spacy.load(app.conf['EL_MODEL'])
nb_model = spacy.load(app.conf['NB_MODEL'])
lt_model = spacy.load(app.conf['LT_MODEL'])```